File: python/src/test_environment.py
#!/usr/bin/env python3
# laurent.martin.aspera@fr.ibm.com
# Common library for sample scripts
# Helper methods to get API environment according to config file
# Simplified function to start transfer and wait for it to finish
import os
import sys
import yaml
import json
import time
import grpc
import logging
import tempfile
import subprocess
import base64
from http.client import HTTPConnection
from urllib.parse import urlparse


# get project root folder
TOP_FOLDER = os.path.abspath(os.path.join(
    os.path.dirname(__file__), '..', '..'))

# config file with sub-paths in project's root folder
PATHS_FILE = 'config/paths.yaml'

# read project's relative paths config file
PATHS = yaml.load(
    open(os.path.join(TOP_FOLDER, *PATHS_FILE.split('/'))), Loader=yaml.FullLoader
)

# Error hint to help user to fix the issue
ERROR_HINT = f'\nPlease check: SDK installed in {PATHS["sdk_root"]}, configuration file: {PATHS["main_config"]}'


def get_path(name):
    '''Get con figuration sub-path in project's root folder'''
    item_path = os.path.join(TOP_FOLDER, *PATHS[name].split('/'))
    assert os.path.exists(item_path), f'ERROR: {item_path} not found.{ERROR_HINT}'
    return item_path


# Read configuration from configuration file
CONFIG = yaml.load(open(get_path('main_config')), Loader=yaml.FullLoader)

# location of gRPC stubs
python_stub_folder = os.path.join(
    get_path('trsdk_noarch'), 'connectors', 'python')

assert os.path.exists(
    python_stub_folder
), f'ERROR: python stubs not found in: {python_stub_folder}.{ERROR_HINT}'

# tell where to find gRPC stubs: transfer_pb2 and transfer_pb2_grpc
sys.path.insert(1, python_stub_folder)

# before stub import: protobuf: avoid incompatibility of version, use pure python implementation
os.environ['PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION'] = 'python'

# import gRPC stubs (Transfer SDK API)
import transfer_pb2_grpc as transfer_manager_grpc  # noqa: E4
import transfer_pb2 as transfer_manager  # noqa: E4

# folder with executables
arch_folder = os.path.join(get_path('sdk_root'), CONFIG['misc']['system_type'])
assert os.path.exists(
    arch_folder
), f'ERROR: SDK not found in: {arch_folder}.{ERROR_HINT}'
# use 'ascp' in PATH, add the one from SDK
os.environ['PATH'] += arch_folder

# set logger for debugging
logging.basicConfig()
logging.getLogger().setLevel(logging.DEBUG)

# debug http: see: https://stackoverflow.com/questions/10588644
HTTPConnection.debuglevel = 1
requests_log = logging.getLogger('requests.packages.urllib3')
requests_log.setLevel(logging.DEBUG)
requests_log.propagate = True

# Global vars
transfer_daemon_process = None
sdk_client = None
shutdown_after_transfer = True
file_list = sys.argv[1:]

assert file_list, f'ERROR: Usage: {sys.argv[0]} <files to send>'

TRANSFER_SDK_DAEMON = 'asperatransferd'


def get_configuration(key):
    '''Get configuration value for specific app'''
    assert key in CONFIG, f'configuration for {key} is missing'
    return CONFIG[key]


def start_daemon(sdk_grpc_url):
    '''
    Start transfer manager daemon if not already running

    @return gRPC client
    '''
    global transfer_daemon_process
    global sdk_client
    # avoid message: 'Other threads are currently calling into gRPC, skipping fork() handlers'
    os.environ['GRPC_ENABLE_FORK_SUPPORT'] = 'false'
    # create a connection to the transfer manager daemon, in case it is running
    grpc_url = urlparse(sdk_grpc_url)
    channel = grpc.insecure_channel(f'{grpc_url.hostname}:{grpc_url.port}')
    # try to start daemon a few times if needed
    for i in range(0, 2):
        try:
            print(
                f'Connecting to {TRANSFER_SDK_DAEMON} using gRPC: {grpc_url.hostname} {grpc_url.port}...'
            )
            grpc.channel_ready_future(channel).result(timeout=3)
            print('SUCCESS: connected')
            # channel is ok, let's get the stub
            sdk_client = transfer_manager_grpc.TransferServiceStub(channel)
        except grpc.FutureTimeoutError:
            print('ERROR: Failed to connect\nStarting daemon...')
            # else prepare config and start
            bin_folder = arch_folder
            log_folder = tempfile.gettempdir()
            # see https://developer.ibm.com/apis/catalog/aspera--aspera-transfer-sdk/Configuration%20File
            config = {
                'address': grpc_url.hostname,
                'port': grpc_url.port,
                'log_directory': log_folder,
                'log_level': 'debug',
                'fasp_runtime': {
                    'use_embedded': False,
                    'user_defined': {
                        'bin': bin_folder,
                        'etc': get_path('trsdk_noarch'),
                    },
                    'log': {
                        'dir': log_folder,
                        'level': 2,
                    },
                },
            }
            tmp_file_base = os.path.join(log_folder, 'daemon')
            # dynamically create a config file
            conf_file = f'{tmp_file_base}.conf'
            with open(conf_file, 'w') as the_file:
                the_file.write(json.dumps(config))
            command = [
                os.path.join(bin_folder, TRANSFER_SDK_DAEMON),
                '--config',
                conf_file,
            ]
            out_file = f'{tmp_file_base}.out'
            err_file = f'{tmp_file_base}.err'
            time.sleep(1)
            print(f'Starting: {" ".join(command)}')
            print(f'stderr: {err_file}')
            print(f'stdout: {out_file}')
            print(f'sdk log: {log_folder}/asperatransferd.log')
            print(f'xfer log: {log_folder}/aspera-scp-transfer.log')
            transfer_daemon_process = subprocess.Popen(
                ' '.join(command),
                shell=True,
                stdout=open(out_file, 'w'),
                stderr=open(err_file, 'w'),
            )
            transfer_daemon_process.poll()
            # give time for startup
            time.sleep(5)
        if sdk_client is not None:
            break
    if sdk_client is None:
        print(
            'ERROR: daemon not started or cannot be started.\nCheck the logs: daemon.err and daemon.out (see paths above).'
        )
        exit(1)
    return sdk_client


def start_transfer(transfer_spec):
    '''Start a transfer and return transfer id'''
    global sdk_client
    # create a transfer request
    transfer_request = transfer_manager.TransferRequest(
        transferType=transfer_manager.FILE_REGULAR,
        config=transfer_manager.TransferConfig(),
        transferSpec=json.dumps(transfer_spec),
    )
    # send start transfer request to transfer manager daemon
    transfer_response = sdk_client.StartTransfer(transfer_request)
    if 4 == transfer_response.status:
        print('ERROR: {0}'.format(transfer_response.error.description))
        exit(1)
    return transfer_response.transferId


def wait_transfer(transfer_id):
    '''Wait for transfer to finish'''
    global sdk_client
    print('transfer started with id {0}'.format(transfer_id))
    # monitor transfer status
    for transfer_info in sdk_client.MonitorTransfers(
        transfer_manager.RegistrationRequest(
            filters=[transfer_manager.RegistrationFilter(
                transferId=[transfer_id])]
        )
    ):
        print('>>>>>>>>>>>>>>>>>>>>>>>>>>>>\ntransfer info {0}'.format(
            transfer_info))
        # check transfer status in response, and exit if it's done
        status = transfer_info.status
        # exit on first success or failure
        if status == transfer_manager.COMPLETED:
            print('finished {0}'.format(status))
            break
        if status == transfer_manager.FAILED:
            raise Exception(transfer_info.message)


def shutdown():
    '''Shutdown transfer manager daemon, if needed'''
    global transfer_daemon_process
    if transfer_daemon_process is not None:
        # transfer_daemon_process.send_signal(signal.CTRL_C_EVENT)
        # transfer_daemon_process.terminate()
        transfer_daemon_process.kill()
        transfer_daemon_process.wait()
        transfer_daemon_process = None
        print('transfer daemon has been terminated')
    else:
        print('transfer daemon not started by this process, or already terminated')


def start_transfer_and_wait(t_spec):
    '''One-call simplified procedure to start daemon, transfer and wait for it to finish'''
    global sdk_client
    global shutdown_after_transfer
    # TODO: remove when transfer sdk bug fixed
    t_spec['http_fallback'] = False
    logging.debug(t_spec)
    try:
        if sdk_client is None:
            sdk_client = start_daemon(CONFIG['misc']['trsdk_url'])
        t_id = start_transfer(t_spec)
        wait_transfer(t_id)
    finally:
        if shutdown_after_transfer:
            shutdown()


def basic_authorization(username, password):
    '''Create basic auth header'''
    return f'Basic {base64.b64encode(f"{username}:{password}".encode()).decode()}'


def basic_auth_header_key_value(username, password):
    '''Create basic auth header key and value for transfer SDK'''
    return {
        'key': 'Authorization',
        'value': basic_authorization(username, password),
    }

File: js/test_environment.js
const fs = require('fs')
const path = require('path')
const yaml = require('js-yaml')
const grpc = require("@grpc/grpc-js")
const protoLoader = require("@grpc/proto-loader")
const assert = require('assert')
const os = require('os')
const { spawn } = require('child_process')

const top_folder = path.resolve(path.dirname(__filename), '..')
const paths_file = "config/paths.yaml"
const paths = yaml.load(fs.readFileSync(path.join(top_folder, paths_file), 'utf8'))
function get_path(name) {
	return path.join(top_folder, paths[name])
}
// read config for examples
const config = yaml.load(fs.readFileSync(get_path("mainconfig"), 'utf8'))
const arch_folder = path.join(get_path("sdk_root"), config["misc"]["system_type"])
const daemon_name = "asperatransferd"
const daemon_exe = path.join(arch_folder, daemon_name)

// load gRPC package definition for aspera transfer sdk
const packageDefinition = protoLoader.loadSync(
	get_path("proto"),
	{ keepCase: true, longs: String, enums: String, defaults: true, oneofs: true })
// get stubs for transfer sdk
const transfersdk = grpc.loadPackageDefinition(packageDefinition).transfersdk

var client = null
var sdk_process = null


module.exports = {
	config: config,
	tmp_folder: os.tmpdir(),
	connect_to_api: (ready_rb) => {
		const grpc_url = new URL(config['misc']['trsdk_url'])
		assert(grpc_url.protocol === 'grpc:', "Expecting gRPC protocol")
		const daemon_conf = {
			address: grpc_url.hostname,
			port: parseInt(grpc_url.port),
			log_directory: os.tmpdir(),
			log_level: 'debug',
			fasp_runtime: {
				use_embedded: false,
				user_defined: {
					bin: arch_folder,
					etc: get_path('trsdk_noarch'),
				},
				log: {
					dir: os.tmpdir(),
					level: 0,
				},
			},
		};
		const tmp_file_base = path.join(os.tmpdir(), 'daemon');
		const conf_file = path.join(get_path("tmpgen"),'daemon.json');
		fs.writeFileSync(conf_file, JSON.stringify(daemon_conf));
		const args = ['-c', conf_file]
		out_file = tmp_file_base + '.out'
		err_file = tmp_file_base + '.err'
		console.log(`Starting: ${daemon_exe} ${args.join(' ')}`)
		console.log(`stderr: ${err_file}`)
		console.log(`stdout: ${out_file}`)
		console.log(`ascp log: ${daemon_conf['log_directory']}/aspera-scp-transfer.log`)
		sdk_process = spawn(daemon_exe, args, {
			stdio: ['ignore', fs.openSync(out_file, 'w'), fs.openSync(err_file, 'w')],
		})
		sdk_process.on('error', (error) => { console.error(`Error starting the child process: ${error.message}`) })
		sdk_process.on('exit', (code, signal) => {
			console.log(`transferd exited (${code})`)
			if (code != 0) {
				throw "transferd exited abnormally"
			}
		})
		console.log(`Started ${daemon_name} with pid ${sdk_process.pid}`)
		// create a connection to the transfer manager daemon
		setTimeout(() => {
			client = new transfersdk.TransferService(grpc_url.hostname + ":" + grpc_url.port, grpc.credentials.createInsecure())
			// try connection, allow 5 seconds
			client.waitForReady((new Date()).getTime() + 5000, (err) => {
				if (err) {
					console.log("No server found...")
					throw err
				}
				console.log("Connected...")
			})
			ready_rb()
		}, 5000)
	},
	shutdown_api: (ok_cb) => {
		console.log("Stopping transferd...")
		sdk_process.on('exit', (code, signal) => { ok_cb() })
		sdk_process.kill(2)
	},
	// start a transfer , provide transfer_spec and optionally event callback
	start_transfer_and_wait: (transferSpec, success_cb) => {
		const startTransferRequest = {
			transferType: 'FILE_REGULAR',
			transferSpec: JSON.stringify(transferSpec)
		}
		const eventStream = client.startTransferWithMonitor(startTransferRequest, function (err, data) {
			console.log("error starting transfer " + err)
			throw err
		})
		eventStream.on('data', function (data) {
			console.log("Transfer %d Mbps/%d Mbps %s %s %s", data.transferInfo.averageRateKbps / 1000,
				data.transferInfo.targetRateKbps / 1000, data.transferEvent, data.status, data.transferType)
			if (data.transferEvent === 'SESSION_STOP' && data.status === 'COMPLETED')
				success_cb()
			if (data.transferEvent === 'SESSION_ERROR' && data.status === 'FAILED')
				throw "ERROR: An error occurred during transfer session"
		})
	}
}
